File: algorithms/0072/solution.cpp
class Solution {
public:
  int minDistance(string word1, string word2) {
    int m = word1.size(), n = word2.size();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    // 单独处理空字符串的情况（边界条件）
    for (int i = 1; i <= m; ++i) {
      dp[i][0] = i;
    }
    for (int j = 1; j <= n; ++j) {
      dp[0][j] = j;
    }

    for (int i = 1; i <= m; ++i) {
      for (int j = 1; j <= n; ++j) {
        dp[i][j] = min(dp[i][j - 1], dp[i - 1][j]) + 1;
        dp[i][j] =
            min(dp[i][j], dp[i - 1][j - 1] + (word1[i - 1] != word2[j - 1]));
      }
    }
    return dp[m][n];
  }
};

File: algorithms/0072/solution.go
func minDistance(word1 string, word2 string) int {
    m, n := len(word1), len(word2)
    // 我们需要使用一个temp数组保存上一轮迭代的状态
    dp, temp := make([]int, m + 1), make([]int, m + 1)
    for i := 1; i <= m; i++ {
        dp[i] = i
        temp[i] = i
    }
    for j := 1; j <= n; j++ {
        dp[0] = j
        for i := 1; i <= m; i++ {
            dp[i] = min(dp[i], dp[i-1]) + 1
            if (word1[i-1] == word2[j-1]) {
                dp[i] = min(dp[i], temp[i-1])
            } else {
                dp[i] = min(dp[i], temp[i-1] + 1)
            }
        }
        for i := 0; i <= m; i++ {
            temp[i] = dp[i]
        } 
    }
    return dp[m]
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}